src/grpc/stream_infer_handler.cc

```diff
@@ -359,13 +359,31 @@ ModelStreamInferHandler::Process(InferHandler::State* state, bool rpc_ok)
       response->mutable_infer_response()->Clear();
       // repopulate the id so that client knows which request failed.
       response->mutable_infer_response()->set_id(request.id());
-      state->step_ = Steps::WRITEREADY;
       if (!state->is_decoupled_) {
+        state->step_ = Steps::WRITEREADY;
         state->context_->WriteResponseIfReady(state);
       } else {
-        state->response_queue_->MarkNextResponseComplete();
-        state->complete_ = true;
-        state->context_->PutTaskBackToQueue(state);
+        InferHandler::State* writing_state = nullptr;
+        std::lock_guard<std::recursive_mutex> lk1(state->context_->mu_);
+        {
+          std::lock_guard<std::recursive_mutex> lk2(state->step_mtx_);
+          bool has_prev_ready_response =
+              state->response_queue_->HasReadyResponse();
+          state->response_queue_->MarkNextResponseComplete();
+          if (!has_prev_ready_response) {
```

Comment:
Why add this condition? This code is executed when InferAsync call returns an error. The `has_prev_ready_response` should always be false in this case right? As there should be no other responses.

Reply:
I see. The InferAsync only return errors up to scheduler enqueue. Yes, `has_prev_ready_response` should always be false.

Updated: [No need to check for previous response on InferAsync error](https://github.com/triton-inference-server/server/pull/7404/commits/4f10e06b3634660b81c29ccbc062e1e500029797)